src/CoreWCF.Primitives/src/CoreWCF/Security/SecuritySessionServerSettings.cs

```diff
@@ -1954,6 +1965,459 @@ public Task OpenAsync(CancellationToken token)
+            bool isInputClosed;
+            bool isOutputClosed;
+            bool sentClose;
+            bool receivedClose;
+            RequestContext closeRequestContext;
+            Message closeResponseMessage;
+
+            public ServerSecurityDuplexSessionChannel(
+                SecuritySessionServerSettings settings,
+               // IServerReliableChannelBinder channelBinder,
+                SecurityContextSecurityToken sessionToken,
+                object listenerSecurityState, SecurityListenerSettingsLifetimeManager settingsLifetimeManager, EndpointAddress address)
+                : base(settings,
+                     // channelBinder,
+                      sessionToken, listenerSecurityState, settingsLifetimeManager, address)
+            {
+                this.session = new SoapSecurityServerDuplexSession(sessionToken, settings, this);
+            }
+
+            public EndpointAddress RemoteAddress
+            {
+                get
+                {
+                    return LocalAddress;
+                }
+            }
+
+            public Uri Via
+            {
+                get
+                {
+                    return this.RemoteAddress.Uri;
+                }
+            }
+
+            public IDuplexSession Session
+            {
+                get
+                {
+                    return this.session;
+                }
+            }
+
+            protected override void AbortCore()
+            {
+                base.AbortCore();
+                this.Settings.RemoveSessionChannel(this.session.Id);
+                CleanupPendingCloseState();
+            }
+
+            void CleanupPendingCloseState()
+            {
+                lock (LocalLock)
+                {
+                    if (this.closeResponseMessage != null)
+                    {
+                        this.closeResponseMessage.Close();
+                        this.closeResponseMessage = null;
+                    }
+                    if (this.closeRequestContext != null)
+                    {
+                        this.closeRequestContext.Abort();
+                        this.closeRequestContext = null;
+                    }
+                }
+            }
+
+            public void Abort()
+            {
+                AbortCore();
+            }
+
+            public Task CloseAsync()
+            {
+                return CloseAsync(new TimeoutHelper(ServiceDefaults.CloseTimeout).GetCancellationToken());
+            }
+
+            public Task CloseAsync(CancellationToken token)
+            {
+                // step 1: close output session
+                this.CloseOutputSession(token);
+
+                // if the channel was aborted while closing the output session, return
+                if (this.State == CommunicationState.Closed)
+                {
+                    return Task.CompletedTask;
+                }
+
+                //Check with Matt about Inp/Oout Session
+
```

Comment:
I believe this is needed

Reply:
Can you elaborate this case/situation little more ? Or we can talk on sync.. I am assuming this is not needed further..